core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java

```diff
@@ -181,21 +183,27 @@ private AggregationAnalyzer(
         // No defensive copy here for performance reasons.
         // Copying this map may lead to quadratic time complexity
         this.columnReferences = analysis.getColumnReferenceFields();
+    }
 
-        // Normal iteration instead of stream for performance reasons
-        ImmutableSet.Builder<FieldId> groupingFieldsBuilder = ImmutableSet.builder();
-        for (Expression expression : groupByExpressions) {
-            ResolvedField resolvedField = columnReferences.get(NodeRef.of(expression));
-            if (resolvedField != null) {
-                groupingFieldsBuilder.add(resolvedField.getFieldId());
+    private Set<FieldId> getGroupingFields()
+    {
+        if (groupingFields == null) {
+            // Normal iteration instead of stream for performance reasons
+            ImmutableSet.Builder<FieldId> groupingFieldsBuilder = ImmutableSet.builder();
+            for (Expression expression : groupByExpressions) {
+                ResolvedField resolvedField = columnReferences.get(NodeRef.of(expression));
+                if (resolvedField != null) {
+                    groupingFieldsBuilder.add(resolvedField.getFieldId());
+                }
             }
-        }
-        this.groupingFields = groupingFieldsBuilder.build();
+            this.groupingFields = groupingFieldsBuilder.build();
 
-        this.groupingFields.forEach(fieldId -> {
-            checkState(isFieldFromScope(fieldId, sourceScope),
-                    "Grouping field %s should originate from %s", fieldId, sourceScope.getRelationType());
-        });
+            this.groupingFields.forEach(fieldId -> {
+                checkState(isFieldFromScope(fieldId, sourceScope),
```

Comment:
But now we are applying this check lazily, would it fail at a different place post `Analysis` phase or a with a different error.

Reply:
This is a sanity check related strictly to this calculations, so it will not fire unless `groupingFields` fields is actually accessed.
And it will never fire unless somebody screws up the code is some other place.

Reply:
That last sentence turned out to be false